File: xposed-stub/src/main/java/de/robv/android/xposed/XC_MethodReplacement.java
package de.robv.android.xposed;

public abstract class XC_MethodReplacement extends XC_MethodHook {

    @Override
    protected final void beforeHookedMethod(MethodHookParam param) throws Throwable {
        param.setResult(replaceHookedMethod(param));
    }

    @Override
    protected final void afterHookedMethod(MethodHookParam param) {
        // no-op
    }

    protected abstract Object replaceHookedMethod(MethodHookParam param) throws Throwable;

    public static XC_MethodReplacement returnConstant(final Object result) {
        return new XC_MethodReplacement() {
            @Override
            protected Object replaceHookedMethod(MethodHookParam param) {
                return result;
            }
        };
    }
}

File: app/src/main/java/io/github/magisk317/relay/xp/compat/XC_MethodReplacement.kt
package io.github.magisk317.relay.xp.compat

abstract class XC_MethodReplacement : XC_MethodHook() {

    final override fun beforeHookedMethod(param: MethodHookParam) {
        param.setResult(replaceHookedMethod(param))
    }

    final override fun afterHookedMethod(param: MethodHookParam) {
        // no-op
    }

    protected abstract fun replaceHookedMethod(param: MethodHookParam): Any?

    companion object {
        fun returnConstant(result: Any?): XC_MethodReplacement {
            return object : XC_MethodReplacement() {
                override fun replaceHookedMethod(param: MethodHookParam): Any? = result
            }
        }
    }
}
